ui/src/views/ThirdPartyLogin/index.jsx

```diff
@@ -103,6 +106,9 @@ export default class ThirdPartyLogin extends Component {
           />
         )}
         {!user && <Homepage />}
+        {!user &&
+          isOneLoginStrategy &&
+          window.open(`/login/${window.env.UI_LOGIN_STRATEGY_NAMES}`, '_blank')}
```

Comment:
This more or less seems to work. Most of the times, I'm seeing two of these pages pop up. It's probably because the redirect to this page is opening a new one since for a split second, `user` is not defined. We probably need to find a way to not open a new window if this view was rendered after a redirect. Perhaps by reading `window.opener`.

```diff
diff --git a/ui/src/views/ThirdPartyLogin/index.jsx b/ui/src/views/ThirdPartyLogin/index.jsx
index 28ef04474..6a4f8486f 100644
--- a/ui/src/views/ThirdPartyLogin/index.jsx
+++ b/ui/src/views/ThirdPartyLogin/index.jsx
@@ -107,7 +107,7 @@ export default class ThirdPartyLogin extends Component {
         )}
         {!user && <Homepage />}
         {!user &&
-          isOneLoginStrategy &&
+          isOneLoginStrategy && window.opener &&
           window.open(`/login/${window.env.UI_LOGIN_STRATEGY_NAMES}`, '_blank')}
       </Dashboard>
     );
```

Reply:
Also, if you can rebase on master, that'd be good :)

Reply:
A relevant changelog explaining the change may be useful. https://github.com/taskcluster/taskcluster/blob/master/dev-docs/best-practices/changelog.md

Reply:
Ugh, I'm afraid `window.opener` won't always work. Sorry, I should have seen this earlier. The good news is that I think we could rely on session storage. I tested the diff below and it seems to work :)

```diff
diff --git a/ui/src/App/Main.jsx b/ui/src/App/Main.jsx
index c48e48039..802d2a2e7 100644
--- a/ui/src/App/Main.jsx
+++ b/ui/src/App/Main.jsx
@@ -66,6 +66,22 @@ export default class Main extends Component {
       fetchPolicy: 'network-only',
     });
     const thirdPartyLogin = isThirdPartyLogin();
+    const isOneLoginStrategy =
+      window.env.UI_LOGIN_STRATEGY_NAMES &&
+      window.env.UI_LOGIN_STRATEGY_NAMES.split(' ').length === 1;
+    const THIRD_PARTY_DID_AUTO_LOGIN_KEY = 'third-party-did-auto-login';
+
+    if (
+      !user &&
+      thirdPartyLogin &&
+      isOneLoginStrategy &&
+      !sessionStorage.getItem(THIRD_PARTY_DID_AUTO_LOGIN_KEY)
+    ) {
+      sessionStorage.setItem(THIRD_PARTY_DID_AUTO_LOGIN_KEY, 'true');
+      window.open(`/login/${window.env.UI_LOGIN_STRATEGY_NAMES}`);
+
+      return;
+    }
 
     if (
       user &&
@@ -74,12 +90,16 @@ export default class Main extends Component {
       data.isLoggedIn === false
     ) {
       onUnauthorize();
+
+      return;
     }
 
     // Users who were logged in earlier manually will be logged out in order to
     // be able to complete the third party login flow.
     if (user && user.identityProviderId === 'manual' && thirdPartyLogin) {
       onUnauthorize();
+
+      return;
     }
 
     // If a third party tries to login but the user is
diff --git a/ui/src/App/index.jsx b/ui/src/App/index.jsx
index 4951988e4..8024efee6 100644
--- a/ui/src/App/index.jsx
+++ b/ui/src/App/index.jsx
@@ -211,7 +211,11 @@ export default class App extends Component {
             <MuiThemeProvider theme={theme}>
               <FontStager />
               <CssBaseline />
-              <Main key={auth.user} routes={routes} error={error} />
+              <Main
+                key={JSON.stringify(auth.user)}
+                routes={routes}
+                error={error}
+              />
             </MuiThemeProvider>
           </ToggleThemeContext.Provider>
         </AuthContext.Provider>
diff --git a/ui/src/views/ThirdPartyLogin/index.jsx b/ui/src/views/ThirdPartyLogin/index.jsx
index 89f77a378..9d9f2d4fa 100644
--- a/ui/src/views/ThirdPartyLogin/index.jsx
+++ b/ui/src/views/ThirdPartyLogin/index.jsx
@@ -35,7 +35,7 @@ export default class ThirdPartyLogin extends Component {
       location: { search },
     } = props;
     const query = parse(search.slice(1));
-    const scopes = query.scope.split(' ');
+    const scopes = (query.scope || []).split(' ');
     const registeredClientId = query.client_id;
 
     if (
@@ -87,9 +87,6 @@ export default class ThirdPartyLogin extends Component {
   render() {
     const { user, data } = this.props;
     const { formData } = this.state;
-    const isOneLoginStrategy =
-      window.env.UI_LOGIN_STRATEGY_NAMES &&
-      window.env.UI_LOGIN_STRATEGY_NAMES.split(' ').length === 1;
 
     return (
       <Dashboard title="Third Party Login">
@@ -106,10 +103,6 @@ export default class ThirdPartyLogin extends Component {
           />
         )}
         {!user && <Homepage />}
-        {!user &&
-          isOneLoginStrategy &&
-          window.opener &&
-          window.open(`/login/${window.env.UI_LOGIN_STRATEGY_NAMES}`, '_blank')}
       </Dashboard>
     );
   }
```